gcp-pubsub-http-connector/main.go

```diff
@@ -0,0 +1,183 @@
+	"context"
+	"errors"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"sync"
+
+	"cloud.google.com/go/pubsub"
+	"go.uber.org/zap"
+	"google.golang.org/api/option"
+
+	"github.com/fission/keda-connectors/common"
+)
+
+type pubsubConnector struct {
+	pubsubInfo    GCPPubsubConnInfo
+	connectordata common.ConnectorMetadata
+	logger        *zap.Logger
+}
+
+// GCPPubsubConnInfo contains the fields needed to connect to the GCP's pub sub topic queue.
+type GCPPubsubConnInfo struct {
+	ProjectID      string
+	SubscriptionID string
+	// Creds points to the json value(not file) which has the service account key
+	Creds string
+}
+
+func main() {
+	logger, err := zap.NewProduction()
+	if err != nil {
+		log.Fatalf("can't initialize zap logger: %v", err)
+	}
+	defer logger.Sync()
+
+	connectordata, err := common.ParseConnectorMetadata()
+	if err != nil {
+		logger.Error("Environment variable is missing ", zap.Error(err))
+		os.Exit(127)
+	}
+
+	pubsubInfo, err := GetGCPInfo()
+	if err != nil {
+		logger.Error("failed to find GCP creds or project name or subscription name", zap.Error(err))
+		return
+	}
+	conn := pubsubConnector{
+		pubsubInfo:    *pubsubInfo,
+		connectordata: connectordata,
+		logger:        logger,
+	}
+
+	logger.Info("Conn: %s", zap.String("Response topic", conn.connectordata.ResponseTopic))
+	conn.consumeMessage()
+}
+
+func (conn pubsubConnector) consumeMessage() {
+
+	creds := []byte(conn.pubsubInfo.Creds)
+	headers := http.Header{
+		"KEDA-Topic":          {conn.connectordata.Topic},
+		"KEDA-Response-Topic": {conn.connectordata.ResponseTopic},
+		"KEDA-Error-Topic":    {conn.connectordata.ErrorTopic},
+		"Content-Type":        {conn.connectordata.ContentType},
+		"KEDA-Source-Name":    {conn.connectordata.SourceName},
+	}
+	ctx := context.Background()
+	client, err := pubsub.NewClient(ctx, conn.pubsubInfo.ProjectID, option.WithCredentialsJSON(creds))
+	if err != nil {
+		conn.logger.Error(err.Error())
+	}
+
+	var mu sync.Mutex
+	sub := client.Subscription(conn.pubsubInfo.SubscriptionID)
+	cctx, _ := context.WithCancel(ctx)
+
+	err = sub.Receive(cctx, func(ctx context.Context, msg *pubsub.Message) {
+		mu.Lock()
+		defer mu.Unlock()
+
+		// Set the attributes as message header came from PubSub record
+		for k, v := range msg.Attributes {
+			headers.Add(k, v)
+		}
+
+		// Push the message to the endpoint
+		resp, err := common.HandleHTTPRequest(string(msg.Data), headers, conn.connectordata, conn.logger)
+		if err != nil {
+			if conn.connectordata.ErrorTopic != "" {
+				conn.responseOrErrorHandler(conn.connectordata.ErrorTopic, "Error", headers)
+			}
+			conn.logger.Error("Error sending the message to the endpoint %v", zap.Error(err))
+		} else {
+			body, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				if conn.connectordata.ErrorTopic != "" {
+					conn.responseOrErrorHandler(conn.connectordata.ErrorTopic, string(body), headers)
+				}
+				conn.logger.Error("Error reading body", zap.Error(err))
+				resp.Body.Close()
```

Comment:
Can we have body close in defer statement before we open like at line 183 like `defer resp.Body.Close()`